activesupport/lib/active_support/callbacks.rb

```diff
@@ -843,7 +843,11 @@ def get_callbacks(name) # :nodoc:
           end
 
           def set_callbacks(name, callbacks) # :nodoc:
-            self.__callbacks = __callbacks.merge(name.to_sym => callbacks)
```

Comment:
I'm curious why we even do the `merge` in the first place versus just assigning, doesn't `__*` imply private enough and that we own it?

Reply:
If `__callbacks=` is never called, `__callbacks` returns super class' `__callbacks`, just assigning will accidentally mutate super class' variable.

https://github.com/rails/rails/blob/bfc5de98871ac5f67d94ff7f379eb06632bef088/activesupport/lib/active_support/core_ext/class/attribute.rb#L101-L114

Reply:
This pattern occurs in the Rails code base in about a half-dozen places.  It's potentially a subtle source of bugs too.  If a parent class' Hash attribute is modified after a subclass modifies its own copy, the subclass will never "inherit" those modifications.

I changed two of these instances in https://github.com/rails/rails/pull/38871/commits/442f091580e1b6e1de19d8e393c3d519409eede7 and https://github.com/rails/rails/pull/38871/commits/e89a747164b501f4c8930984a3df1c2d782495e6 (neither merged yet), but I've been wondering if there's some pattern we can abstract.  Like a specialized `class_hash` macro instead of `class_attribute`, or a specialized `InheritingHash` type to use with `class_attribute`.

Reply:
If that is the bahavior you want I don't think `class_attribute` is what you need. A regular method that call `super` would have the behavior that you want.

I don't think it warrants a new abstraction.

Reply:
But I'd say I entertain a PR if you have an implementation in mind to those new patterns. If you could also change the places in Rails that use it in the same PR it would be a good example of its usefulness.